File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        
        phone = {"2": "abc", "3": "def", "4": "ghi", "5": "jkl", "6": "mno", "7": "pqrs", "8": "tuv", "9": "wxyz"}
        res = []
        
        def backtrack(combination, next_digits):
            if not next_digits:
                res.append(combination)
                return
            
            for letter in phone[next_digits[0]]:
                backtrack(combination + letter, next_digits[1:])
        
        backtrack("", digits)
        return res

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> s = new ArrayList<>();
        
        if(digits == null || digits.length() == 0)
            return s;
        
        int len = digits.length();
        int arr[] = new int[len];
        for(int i =0 ;i< len ; i++){
            arr[i]  = digits.charAt(i)-'0';
        }
        
        String letters[] = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        Queue<String> que = new LinkedList<>();
        que.offer(""); //better then queue.add() as it does not throw run time exception
        
        for(int i =0 ; i< len; i++){
            String word = letters[arr[i]-2];
            int size = que.size();
            for (int j = 0; j < size; j++) {
                String temp = que.poll();
                for (char ch : word.toCharArray()) {
                    que.offer(temp + ch);
                }
            }
        }
        return new ArrayList<>(que);
    }
}